Database access AWS RDS Proxy support

greedy52 opened the pull request:
Implements #9820
- #9820

This is a rework of #9824 which couldn't get merged on time because of other conflicts.

### Background
- [RDS proxy](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-proxy-setup.html) is a RDS service where proxy instances are deployed to access other RDS/Aurora databases. 
- A typical use case: AWS Lambda executions can easily make thousands of new RDS connections per second. RDS Proxies can be used as a middleman that serves all the new connections from lambda but maintains a small connection pool to the backend RDS. 
- Each RDS proxy has one "default" endpoint. Extra [custom endpoints](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-proxy-endpoints.html) can be added to each RDS proxy with a different target role (e.g target Aurora readers)


### UX
RDS Proxy is added as new AWS database type
- Static config: only a valid `uri` is required. no new fields. all info will be parsed from the URI.
- Auto-discovery: new `rdsproxy` matcher. 

### Security
- Teleport ---(RDS IAM Auth)--> RDS Proxies ---(user/password from secrets)--> RDS
- IAM permissions for auto-discovery: `rds:DescribeDBProxies`, `rds:DescribeDBProxyEndpoints`, `rds:DescribeDBProxyTargets`, `rds:ListTagsForResource`

### Leftover TODOs
- #17339


### Testing
```
$ tsh db ls
Name                        Description                        Allowed Users Labels   
--------------------------- ---------------------------------- ------------- ---------------
steve-postgres-proxy        RDS Proxy in ca-central-1          [*]           Env=dev,...
steve-postgres-proxy-myread RDS Proxy endpoint in ca-central-1 [*]           Env=dev,...,endpoint-type=READ_ONLY...
```

- [x] Aurora Postgres
- [x] Aurora MySQL
- [x] RDS MariaDB

greedy52 committed:
Database access AWS RDS Proxy support

greedy52 committed:
fix typo/lint

greedy52 committed:
Merge branch 'master' of github.com:gravitational/teleport into STeve/9820_rds_proxy_support

greedy52 committed:
fix typo

greedy52 committed:
bypass RDS Proxy for SQL server

greedy52 committed:
fix typos

smallinsky commented on lib/srv/db/cloud/mocks.go:
```diff
@@ -133,6 +136,73 @@ func (m *RDSMock) ModifyDBClusterWithContext(ctx aws.Context, input *rds.ModifyD
+		}, nil
+	}
+
+	endpoints := []*rds.DBProxyEndpoint{}
```
nit: 
https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

```suggestion
	var endpoints  []*rds.DBProxyEndpoint
```

smallinsky commented on lib/srv/db/cloud/watchers/rds_proxy.go:
```diff
@@ -0,0 +1,213 @@
+		f.log.Debugf("Failed to get RDS Proxy endpoints: %v.", err)
+	}
+
+	databases := types.Databases{}
```
nit: [CodeReviewComments#declaring-empty-slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices)

```suggestion
	var databases types.Databases
```

smallinsky commented on lib/srv/db/cloud/watchers/rds_proxy.go:
```diff
@@ -0,0 +1,213 @@
+			continue
+		}
+
+		if !services.IsRDSProxyAvailable(dbProxy) {
```
What would you say about adding a Health-ping-check call ensuring that proxy is reachable. This will help avoid the scenario where db agent is deployed outside VPC and has the right permission to discovery RDS proxy but connection will fail because db agent is not able to dial RDS Proxy VPC from outside

greedy52 replied:
I believe this applies to other database types too. I usually deploy Aurora, ElastiCache etc. in private subnets so they are not accessible from outside the VPC. 

I am not sure if "watchers" is the best place to do pings as it can block for a long time and delays other discoveries. I feel maybe better to have some kind of health check for every database after they are "registered". What do you think? A "health-check" feature will benefit non cloud-hosted databases as well.

smallinsky commented on api/types/database.go:
```diff
@@ -680,6 +693,16 @@ func (gcp GCPCloudSQL) GetServerName() string {
 }
 
+// HasValidID returns true if any of the identifiers exists.
+func (rds RDS) HasValidID() bool {
```
```suggestion
func (rds RDS) IsValid() bool {
```
HasValidID at first glance suggest that RDS struct has a ID field that we are validating in the `HasValidID` call.

smallinsky commented on api/types/database.go:
```diff
@@ -680,6 +693,16 @@ func (gcp GCPCloudSQL) GetServerName() string {
+}
+
+// IsRDSProxy returns true if this is an RDS Proxy.
+func (rds RDS) IsRDSProxy() bool {
```
nit: right now we have: 
`RDS.IsRDSProxy() `
what would you say about just: 
`RDS.IsProxy()`
```suggestion
func (rds RDS) IsProxy() bool {
```

smallinsky commented on lib/srv/db/cloud/watchers/watcher.go:
```diff
@@ -167,7 +167,7 @@ func makeFetchers(ctx context.Context, config *WatcherConfig) (result []Fetcher,
 	type makeFetcherFunc func(cloud.Clients, string, types.Labels) (Fetcher, error)
 	makeFetcherFuncs := map[string][]makeFetcherFunc{
-		services.AWSMatcherRDS:         {makeRDSInstanceFetcher, makeRDSAuroraFetcher},
+		services.AWSMatcherRDS:         {makeRDSInstanceFetcher, makeRDSAuroraFetcher, makeRDSProxyFetcher},
```
Wonder if we should consider making the makeRDSProxyFetcher as a separate matcher like `rdsproxy `  and make the discovery flow independed from current `rds` auto discovery flow .

I can imagine a scenario where a customer uses RDS Proxy access with some default tags like '*' 
without enabling IMA auth on RDS instance.  By default teleport agent will show RDS Proxy with RDS instances  where access though RDS instances will be never used.

greedy52 replied:
Yes. Totally agreed. There is no point to have both RDS and RDS Proxy for the same database. I actually started with a separate matcher but removed it at some point. I will add it back.

mdwn reviewed (approved):
I have one small nit, but otherwise I think this LGTM.

greedy52 committed:
refactor to use dedicated "rdsproxy" database type and aws match

greedy52 commented:
@smallinsky as discussed, I've split RDS Proxy as a separate database type and a separate matcher.

There are a lot of files changed but mostly really minor.

I did find a bug in the existing configurator code where we only add IAM permissions for RDS autodiscovery but do not add IAM permissions for static RDS databases. I've addressed that in the same commit.

Retested with MySQL and Postgres.

smallinsky commented:
The `ensureIAMAuth` for rdsproxy  auto-discovery always returns following errors:  
```
 ERRO [IAM]       Failed to auto-configure IAM for Database(Name=marek-mysql-test, Type=rdsproxy, Labels=map[account-id:1234 engine:MYSQL region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-123]). error:[
ERROR REPORT:
Original Error: trace.aggregate no RDS cluster ID or instance ID for Database(Name=marek-mysql-test, Type=rdsproxy, Labels=map[account-id:1234 engine:MYSQL region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-123])
Stack Trace:
	/home/ec2-user/teleport/lib/srv/db/cloud/aws.go:114 github.com/gravitational/teleport/lib/srv/db/cloud.(*awsClient).setupIAM
	/home/ec2-user/teleport/lib/srv/db/cloud/iam.go:273 github.com/gravitational/teleport/lib/srv/db/cloud.(*IAM).processTask
	/home/ec2-user/teleport/lib/srv/db/cloud/iam.go:122 github.com/gravitational/teleport/lib/srv/db/cloud.(*IAM).Start.func1
	/home/ec2-user/go/go1.19.2/src/runtime/asm_amd64.s:1594 runtime.goexi
```
We should handle this error more gracefully.

greedy52 committed:
fix RDS Proxy ensure IAM Auth error

greedy52 commented:
> @greedy52 The `ensureIAMAuth` for rdsproxy auto-discovery always returns following errors:
> 
> ```
>  ERRO [IAM]       Failed to auto-configure IAM for Database(Name=marek-mysql-test, Type=rdsproxy, Labels=map[account-id:1234 engine:MYSQL region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-123]). error:[
> ERROR REPORT:
> Original Error: trace.aggregate no RDS cluster ID or instance ID for Database(Name=marek-mysql-test, Type=rdsproxy, Labels=map[account-id:1234 engine:MYSQL region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-123])

Good catch. It even showed up in unit test. Fixed now in https://github.com/gravitational/teleport/pull/17005/commits/1b345a9c0b45c4ba5411db7fbfc1fd701b0df1ee

greedy52 committed:
Merge branch 'master' of github.com:gravitational/teleport into STeve/9820_rds_proxy_support

greedy52 committed:
add missing IAM permission rds:ListTagsForResource
